cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java

```diff
@@ -183,8 +190,59 @@ protected void configure() {
     }
 
     dag = new ControlDag(plan.getPhaseConnections());
-    // after flattening, there is guaranteed to be just one source
-    dag.flatten();
+    if (plan.getConditionPhaseBranches().isEmpty()) {
+      // after flattening, there is guaranteed to be just one source
+      dag.flatten();
+    } else {
+      Map<String, ConditionBranches> conditionBranches = plan.getConditionPhaseBranches();
+      Set<String> conditions = conditionBranches.keySet();
+      // flatten only the part of the dag starting from sources and ending in conditions/sinks.
+      Set<String> dagNodes = dag.accessibleFrom(dag.getSources(), Sets.union(dag.getSinks(), conditions));
+      Set<String> dagNodesWithoutCondition = Sets.difference(dagNodes, conditions);
+      Dag subDag = dag.createSubDag(dagNodesWithoutCondition);
+      ControlDag cdag = new ControlDag(subDag);
+      cdag.flatten();
+
+      Set<Connection> connections = new HashSet<>();
+      // Add all connections from cdag
+      Deque<String> bfs = new LinkedList<>();
+      bfs.addAll(cdag.getSources());
+      while (bfs.peek() != null) {
+        String node = bfs.poll();
+        for (String output : cdag.getNodeOutputs(node)) {
+          connections.add(new Connection(node, output));
+          bfs.add(output);
+        }
+      }
+
+      // Now add connections starting from condition
+      bfs.addAll(cdag.getSinks());
```

Comment:
instead of putting the sink in bfs, I believe you need to add connections from the sink to all the conditions in the intersection. This is to handle the case where the sink is a newly added join node, for which dag.getNodeOutputs() won't return anything.